src/az_ulib_ustream_forward/az_ulib_ustream_forward.c

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "az_ulib_port.h"
+#include "az_ulib_result.h"
+#include "azure/core/az_span.h"
+#include "az_ulib_ustream_forward.h"
+
+#include <azure/core/internal/az_precondition_internal.h>
+
+#ifdef __clang__
+#define IGNORE_POINTER_TYPE_QUALIFICATION \
+  _Pragma("clang diagnostic push")        \
+      _Pragma("clang diagnostic ignored \"-Wincompatible-pointer-types-discards-qualifiers\"")
+#define IGNORE_MEMCPY_TO_NULL _Pragma("GCC diagnostic push")
+#define IGNORE_CAST_QUALIFICATION \
+_Pragma("clang diagnostic push") _Pragma("clang diagnostic ignored \"-Wcast-qual\"")
+#define RESUME_WARNINGS _Pragma("clang diagnostic pop")
+#elif defined(__GNUC__)
+#define IGNORE_POINTER_TYPE_QUALIFICATION \
+  _Pragma("GCC diagnostic push") _Pragma("GCC diagnostic ignored \"-Wdiscarded-qualifiers\"")
+#define IGNORE_MEMCPY_TO_NULL _Pragma("GCC diagnostic push")
+#define IGNORE_CAST_QUALIFICATION \
+_Pragma("GCC diagnostic push") _Pragma("GCC diagnostic ignored \"-Wcast-qual\"")
+#define RESUME_WARNINGS _Pragma("GCC diagnostic pop")
+#else
+#define IGNORE_POINTER_TYPE_QUALIFICATION __pragma(warning(push));
+#define IGNORE_MEMCPY_TO_NULL \
+  __pragma(warning(push));  \
+  __pragma(warning(suppress: 6387));
+#define IGNORE_CAST_QUALIFICATION __pragma(warning(push));
+#define RESUME_WARNINGS __pragma(warning(pop));
+#endif // __clang__
+
+static az_result concrete_flush(
+    az_ulib_ustream_forward* ustream_forward,
+    az_ulib_flush_callback flush_callback, 
+    az_ulib_callback_context flush_callback_context);
+static az_result concrete_read(
+    az_ulib_ustream_forward* ustream_forward,
+    uint8_t* const buffer,
+    size_t buffer_length,
+    size_t* const size);
+static az_result concrete_get_size(
+    az_ulib_ustream_forward* ustream_forward, 
+    size_t* const size);
+static az_result concrete_dispose(
+    az_ulib_ustream_forward* ustream_forward);
+static const az_ulib_ustream_forward_interface api
+    = { concrete_flush, concrete_read,  concrete_get_size, concrete_dispose };
+
+static az_result concrete_flush(
+    az_ulib_ustream_forward* ustream_forward,
+    az_ulib_flush_callback flush_callback, 
+    az_ulib_callback_context flush_callback_context)
+{
+  // precondition checks
+  _az_PRECONDITION_NOT_NULL(ustream_forward);
+  _az_PRECONDITION(AZ_ULIB_USTREAM_FORWARD_IS_TYPE_OF(ustream_forward, api));
+  _az_PRECONDITION_NOT_NULL(flush_callback);
+  _az_PRECONDITION_NOT_NULL(flush_callback_context);
+
+  AZ_ULIB_TRY
+  {
+    // point to data from instance
+    size_t buffer_size;
+    AZ_ULIB_THROW_IF_AZ_ERROR(concrete_get_size(ustream_forward, &buffer_size));
+    IGNORE_CAST_QUALIFICATION
```

Comment:
>IGNORE_CAST_QUALIFICATION [](http://example.com/codeflow?start=4&length=25)

Not necessary, please remove.

Reply:
it is necessary for linux machines. they raise an exception "cast discards `const` qualifier from pointer target type [-Werror=cast-qual]."